### PR TITLE
SPRINT 3 RELEASE

### PR DESCRIPTION
**Pull request by RafterGit**
## Overview
PR to merge the user stories that were completed in Sprint 3. 

### List of completed user stories
* [Add Github Badges](https://github.com/raft-tech/tanf-app/issues/187)
* [As a new user, I can enter my first and last name](https://github.com/raft-tech/tanf-app/issues/191)
* [Add integration tests for custom OIDC logic](https://github.com/raft-tech/tan)
* [As a user who is not yet logged in, I see a landing page based on the design](https://github.com/raft-tech/tanf-app/issues/172)
* [As a Tadpole, I want clarity on AC, DoD, and story points so I know I'm doing the right work](https://github.com/raft-tech/TANF-app/issues/221)
* [Compile a list of states to learn from and interview for Round 3 (September 2020)](https://github.com/raft-tech/tanf-app/issues/163)
* [Define manual testing practices](https://github.com/raft-tech/TANF-app/issues/173)
* [Document Retro feedback 9/1](https://github.com/raft-tech/TANF-app/issues/243)
* [Finalize user research goals for STT research](https://github.com/raft-tech/TANF-app/issues/226)
* [Get a tour of OFA resource mailbox to determine whether access is needed
](https://github.com/raft-tech/tanf-app/issues/167)
* [Synthesize feedback from research sessions with OFA (August 2020)](https://github.com/raft-tech/TANF-app/issues/113)

### List of user stories moving to next sprint

* [As a product team, we need to understand how users navigate TDP flow](https://github.com/raft-tech/TANF-app/issues/211)
* [Draft expected outcomes for STT research](https://github.com/raft-tech/TANF-app/issues/227)
* [Discuss process to merge PR to main HHS repo](https://github.com/raft-tech/TANF-app/issues/186)

### User research artifacts delivered:
 
* [Round 3 user research goals](https://teams.microsoft.com/l/file/390A256B-F1B3-403D-BA50-06820EEA7C53?tenantId=d58addea-5053-4a80-8499-ba4d944910df&fileType=docx&objectUrl=https%3A%2F%2Fhhsgov.sharepoint.com%2Fsites%2FTANFDataPortalOFA%2FShared%20Documents%2FDesign%2FSTT%20Research%20September%202020%2FSTT%20Research%20Goals%20-%20Round%203.docx&baseUrl=https%3A%2F%2Fhhsgov.sharepoint.com%2Fsites%2FTANFDataPortalOFA&serviceName=teams&threadId=19:e92913e3e7d443adb823e6497dff1fb3@thread.skype&groupId=41f194a6-c1d3-4680-933e-c8ee7d17e287):lock:
* [Criteria for recruiting STTs](https://teams.microsoft.com/l/file/390A256B-F1B3-403D-BA50-06820EEA7C53?tenantId=d58addea-5053-4a80-8499-ba4d944910df&fileType=docx&objectUrl=https%3A%2F%2Fhhsgov.sharepoint.com%2Fsites%2FTANFDataPortalOFA%2FShared%20Documents%2FDesign%2FSTT%20Research%20September%202020%2FSTT%20Research%20Goals%20-%20Round%203.docx&baseUrl=https%3A%2F%2Fhhsgov.sharepoint.com%2Fsites%2FTANFDataPortalOFA&serviceName=teams&threadId=19:e92913e3e7d443adb823e6497dff1fb3@thread.skype&groupId=41f194a6-c1d3-4680-933e-c8ee7d17e287):lock:
* [STTs attributes](https://teams.microsoft.com/l/file/A12B53AF-3CD1-40C3-93CF-AE26DCD51264?tenantId=d58addea-5053-4a80-8499-ba4d944910df&fileType=xlsx&objectUrl=https%3A%2F%2Fhhsgov.sharepoint.com%2Fsites%2FTANFDataPortalOFA%2FShared%20Documents%2FDesign%2FSTT%20Research%20September%202020%2FSTTs%20Attributes.xlsx&baseUrl=https%3A%2F%2Fhhsgov.sharepoint.com%2Fsites%2FTANFDataPortalOFA&serviceName=teams&threadId=19:e92913e3e7d443adb823e6497dff1fb3@thread.skype&groupId=41f194a6-c1d3-4680-933e-c8ee7d17e287):lock:
* [OFA research synthesis](https://teams.microsoft.com/l/file/588709DE-FE1F-49E5-8D52-CBC6E25FA9CD?tenantId=d58addea-5053-4a80-8499-ba4d944910df&fileType=docx&objectUrl=https%3A%2F%2Fhhsgov.sharepoint.com%2Fsites%2FTANFDataPortalOFA-UserResearch%2FShared%20Documents%2FUser%20Research%2FOFA%20Research%20August%202020%2F%23113%20-%20OFA%20Interviews%20Synthesis%20DRAFT%202020.09.10.docx&baseUrl=https%3A%2F%2Fhhsgov.sharepoint.com%2Fsites%2FTANFDataPortalOFA-UserResearch&serviceName=teams&threadId=19:3bb622256c0b4792976dc5f1f4654fca@thread.skype&groupId=41f194a6-c1d3-4680-933e-c8ee7d17e287):lock:
* [OFA research mailbox notes](https://teams.microsoft.com/l/file/2542A110-040B-44C1-B827-06A54C6EC9F2?tenantId=d58addea-5053-4a80-8499-ba4d944910df&fileType=docx&objectUrl=https%3A%2F%2Fhhsgov.sharepoint.com%2Fsites%2FTANFDataPortalOFA-UserResearch%2FShared%20Documents%2FUser%20Research%2FOFA%20Resource%20Mailbox%20notes.docx&baseUrl=https%3A%2F%2Fhhsgov.sharepoint.com%2Fsites%2FTANFDataPortalOFA-UserResearch&serviceName=teams&threadId=19:3bb622256c0b4792976dc5f1f4654fca@thread.skype&groupId=41f194a6-c1d3-4680-933e-c8ee7d17e287):lock:

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,3 +1,16 @@
...
+|**Frontend Coverage**| [![Codecov-Frontend-Dev](https://codecov.io/gh/raft-tech/TANF-app/branch/raft-tdp-main/graph/badge.svg?flag=dev-frontend)](https://codecov.io/gh/raft-tech/TANF-app?flag=dev-frontend) | [![Codeco-Frontend-HHS](https://codecov.io/gh/HHS/TANF-app/branch/main/graph/badge.svg?flag=main-frontend)](https://codecov.io/gh/raft-tech/TANF-app?flag=main-frontend)   |
+|**Backend Coverage**|  [![Codecov-Backend-Dev](https://codecov.io/gh/raft-tech/TANF-app/branch/raft-tdp-main/graph/badge.svg?flag=dev-backend)](https://codecov.io/gh/raft-tech/TANF-app/branch/raft-tdp-main?flag=dev-backend)|   [![Codecov-Backend-HHS]( https://codecov.io/gh/HHS/TANF-app/branch/main/graph/badge.svg?flag=main-backend)](https://codecov.io/gh/raft-tech/TANF-app/branch/main?flag=main-backend) |
+
+***Due to limitations imposed by Github and occasional slow sever response times, some badges may require a page refresh to load.**
```
Comment (shubhi-raft) [inline README.md:12]:
```suggestion
***Due to limitations imposed by Github and occasional slow server response times, some badges may require a page refresh to load.**
```

**Comment by RafterGit** [comment]
@lauraGgit 

Making a note of an issue that was discovered yesterday (09/13/2020) with Cloud Foundry's CLI tool installation that is causing CircleCi to fail at the `deploy` step as shown [here](https://app.circleci.com/pipelines/github/raft-tech/TANF-app/598/workflows/839803d4-5033-4613-bb9e-273d56356092/jobs/1072/parallel-runs/0/steps/0-104).

The [GPG key](https://gnupg.org/) used by Cloud Foundry for this installation process is expired and consequently throws an error preventing the install.

The maintainers of this project are currently tracking the issue here in their GitHub Repository: https://github.com/cloudfoundry/cli/issues/2046 And they've also opened a support ticket with Pivotal: https://www.pivotaltracker.com/n/projects/892938/stories/174788941

We'll monitor when the fix is implemented upstream and update accordingly.

File: tdrs-frontend/src/components/Footer/Footer.jsx [hunk 2]
```diff
@@ -0,0 +1,56 @@
+import React from 'react'
+import { Footer, Logo, SocialLinks } from '@trussworks/react-uswds'
+
+import ACFLogo from '../../assets/placeholder.png'
```
Comment (lauraGgit) [inline tdrs-frontend/src/components/Footer/Footer.jsx:4]: See @cmajel 's comment about adding in the logo.

File: tdrs-backend/tdpservice/users/api/authorization_check.py [hunk 3]
```diff
@@ -20,7 +20,14 @@ def get(self, request, *args, **kwargs):
         """Handle get request and authenticate user."""
         user = request.user
         if user.is_authenticated:
-            auth_params = {"authenticated": True, "user": {"email": user.username}}
+            auth_params = {
+                "authenticated": True,
+                "user": {
```
Comment (lauraGgit) [inline tdrs-backend/tdpservice/users/api/authorization_check.py:25]: I wonder if here you'd be able to take advantage of the [UserSerializer](https://github.com/HHS/TANF-app/pull/55/files#diff-eb126d5c965960aa16268e1579183160R8) and nest it: https://www.django-rest-framework.org/api-guide/serializers/#dealing-with-nested-objects

**Review by lauraGgit — COMMENTED** [review]
# Overview
This sprint looks good.


## Local development
~It looks like the `REACT_APP_BACKEND_URL=http://localhost:8080/v1`is not getting picked up when `docker-compose up -d --build` is run on the frontend. Perhaps an additional step is needed on the frontend readme for local dev?~ Added.

# QASP Deliverables 
## Deliverable 1: Accepted Features
* Add Github Badges
	AC: Complete
* As a new user, I can enter my first and last name
     AC: As a new user, after login I can: complete
     ~However, the UI does not meet the UI mock up~Complete
* Add integration tests for custom OIDC logic
    AC: login.py tested
    ~AC: check on httpCookie question~ migrating into task of [#231](https://github.com/raft-tech/TANF-app/issues/231#issuecomment-692802118)
* As a user who is not yet logged in, I see a landing page based on the design
    AC: When the user is not logged in, they see a landing page that matches the figma design:
    Update: Feedback addressed.
    >  ***feedback***
    >    - The right widths for the login and resources card. This is what I meant by grid 
    >    - I see it's enabled but the col lengths for those elements seem off
    >	-  There are social links in the footer (not in the design)
    >	- The ACF logo is missing from the footer
    >	- The TANF email isn't styled as a link/a link
    AC: This page renders at all form factors specified in the design (aka is "responsive") complete
    AC: Use the text and placeholders from the mockup: Complete
* As a Tadpole, I want clarity on AC, DoD, and story points so I know I'm doing the right work
    AC: complete
* Define manual testing practices
    AC: complete
* Document Retro feedback 9/1
    This story is still in progress.

### Design review - TK
* Finalize user research goals for STT research
* Compile a list of states to learn from and interview for Round 3 (September 2020)
* Get a tour of OFA resource mailbox to determine whether access is needed
* Synthesize feedback from research sessions with OFA (August 2020)
 
## Deliverable 2: Tested Code
* Cypress test updates could better reflect the updated functionality.

## Deliverable 3: Properly Styled code
* Linting looks good

## Deliverable 4: Accessible
* Pa11y tests pass, however, because of the changes requested to implement #172, the full manual accessibility of the site has not been reviewed. 
 
## Deliverable 5: Deployed
*~ Because of updated made outside of the deadline, the CF cli cannot be installed and the latest version is not deployed. [Please see the note above. This is caused by an external issue.](https://github.com/HHS/TANF-app/pull/55#issuecomment-691983241)~
Resolved by cloud foundry

## Deliverable 6: Documented
* Documentation updates look in line with what was delivered. ~Though, I saw that Makedoc is no longer being built, do you think you’ll replace that?~ Comment addressed.
 
## Deliverable 7: Secure
* Vulnerabilities have been addressed or remediated.

## Deliverable 8: User Research
* Looks good.

File: tdrs-backend/tdpservice/users/api/login.py [hunk 4]
```diff
@@ -31,6 +31,61 @@
...
+        if user is not None:
+            login(
+                request,
+                user,
+                backend="tdpservice.users.authentication.CustomAuthentication",
+            )
+            datetime_time = datetime.datetime.fromtimestamp(time.time())
+            logger.info(f"Found User:  {user.username} on {datetime_time}(UTC)")
+        else:
+            User = get_user_model()
```
[the remark this answers is not in the document]
Comment (carltonsmith) [reply tdrs-backend/tdpservice/users/api/login.py:73]: Yes, I can see how the login portion would be DRYer as it's own method. We can get that done, certainly. Was there more than just the login?

[the remark this answers is not in the document]
Comment (RafterGit) [reply tdrs-backend/tdpservice/users/api/login.py:73]: A remediation for this inquiry has been provided in this PR in the raft-tech Repo:

https://github.com/raft-tech/TANF-app/pull/273

File: tdrs-backend/tdpservice/users/test/test_api.py [hunk 5]
```diff
@@ -38,3 +37,159 @@ def test_create_user(api_client, user_data):
...
+def test_set_profile_data_empty_first_name_and_last_name(api_client, user):
+    """Test profile data cannot be set if first and last name are blank."""
+    api_client.login(username=user.username, password="test_password")
+    response = api_client.post(
+        "/v1/users/set_profile/", {"first_name": "", "last_name": ""},
+    )
+    assert response.status_code == status.HTTP_400_BAD_REQUEST
+
+
+@pytest.mark.django_db
```
[the remark this answers is not in the document]
Comment (RafterGit) [reply tdrs-backend/tdpservice/users/test/test_api.py:114]: Expanded test cases have been added to account for unexpected  fields and missing fields in this backend remediation PR.

https://github.com/raft-tech/TANF-app/pull/273

File: tdrs-backend/tdpservice/users/test/test_auth.py [hunk 6]
```diff
@@ -49,49 +59,350 @@ def test_get_non_user(user):
...
 
+
 def test_oidc_logout(api_client):
     """Test logout url redirects."""
     response = api_client.get("/v1/logout/oidc")
     assert response.status_code == status.HTTP_302_FOUND
 
+
 def test_oidc_logout_with_token(api_client):
     """Test logging out with token redirects and token is removed."""
```
[the remark this answers is not in the document]
Comment (RafterGit) [reply tdrs-backend/tdpservice/users/test/test_auth.py:76]: The test cases have been refactored to properly demonstrate their intended functionality in the backend remediation PR.

https://github.com/raft-tech/TANF-app/pull/273

File: tdrs-backend/tdpservice/users/api/authorization_check.py (hunk 3 above)
Comment (RafterGit) [reply tdrs-backend/tdpservice/users/api/authorization_check.py:25]: As discussed during the [09/15/2020 dev pairing session](https://hhsgov.sharepoint.com/sites/TANFDataPortalOFA/_layouts/15/Doc.aspx?OR=teams&action=edit&sourcedoc={22F8187F-7AC2-4198-8673-7C111DDFD4E5})  this was a suggestion. Due to the simple nature of the JSON returned in this response the overhead of implementing the UserSerializer and nested object support would be unnecessary unless the request grows to be more complex.

File: tdrs-frontend/src/components/Footer/Footer.jsx (hunk 2 above)
Comment (RafterGit) [reply tdrs-frontend/src/components/Footer/Footer.jsx:4]: This was addressed in the frontend remediations PR currently up for review in the raft-tech repo.

https://github.com/raft-tech/TANF-app/pull/262

**Comment by RafterGit** [comment]
@lauraGgit 

## Deliverable 6: Documented
---
**Documentation updates look in line with what was delivered. Though, I saw that Makedoc is no longer being built, do you think you’ll replace that?**


We will be replacing this, the Mackdocs integration in the docker-compose configuration was a temporary proof of concept while we worked towards a permanent documentation approach so we can serve it without needing someone to pull down and spin up a local instance of the application such as the items defined in story [265](https://github.com/raft-tech/TANF-app/issues/265).

**Comment by RafterGit** [comment]
@lauraGgit 

## Deliverable 5: Deployed

**Because of updates made outside of the deadline, the CF cli cannot be installed and the latest version is not deployed. Please see the note above. This is caused by an external issue.**

The external issue with the Cloud Foundry CLI tool installation has since been resolved as reported here [Cloud Foundry GPG Key Expired Issue](https://github.com/cloudfoundry/cli/issues/2046)

Upon merging of the remediation PRs, the CI/CD pipeline will have no issue installing and using the `cf-cli` tool, so this will be resolved.

**Comment by carltonsmith** [comment]
## Deliverable 2: Tested Code

**Cypress test updates could better reflect the updated functionality.**

Updates to our cypress integration testing will be fleshed out more, for example leveraging the Cypress OTP tool to facilitate the Login.gov authentication process. 

Details of this have been captured in the [09/15/2020 dev pairing session](https://hhsgov.sharepoint.com/sites/TANFDataPortalOFA/_layouts/15/Doc.aspx?OR=teams&action=edit&sourcedoc={22F8187F-7AC2-4198-8673-7C111DDFD4E5}) and in issue [231](https://github.com/raft-tech/TANF-app/issues/231).

For this PR we've expanded the current cypress test to capture items on the Landing Page which is part of the raft-tech frontend remediation PR [here](https://github.com/raft-tech/TANF-app/pull/262/files#diff-3c2e0cd8168567b089e821afd38dd2f9R14).

**Review by lauraGgit — APPROVED** [review]

**Comment by lauraGgit** [comment]
@iamjolly / @lfrohlich This sprint looks good from me.

**Comment by iamjolly** [comment]
Thanks @lauraGgit. This is also good from my perspective, @lfrohlich.

**Comment by lfrohlich** [comment]
Thanks, all. This looks good from my perspective.

**Comment by lauraGgit** [comment]
Additional notes from teams:
Vulnerabilities - In running snyk locally I saw that "Dependency @fortawesome/fontawesome-svg-core was not found in package-lock.json. Your package.json and package-lock.json are probably out of sync. Please run "npm install" and try again." but run against package.json it passes. This would be great to get remediated next sprint.

Page title accessibility concerns have been captured in [#283](https://github.com/raft-tech/TANF-app/issues/283)

Testing of the HTTPcookie will be added as a task to [#231](https://github.com/raft-tech/TANF-app/issues/231#issuecomment-692802118)

**Pull request merged**